pkg/controller/apiextensions/composite/api.go

```diff
@@ -82,6 +83,27 @@ func (a *APIFilteredSecretPublisher) UnpublishConnection(_ context.Context, _ re
 	return nil
 }
 
+// NewCompositionSelectorChain returns a new CompositionSelectorChain.
+func NewCompositionSelectorChain(list ...CompositionSelector) *CompositionSelectorChain {
+	return &CompositionSelectorChain{list: list}
+}
+
+// CompositionSelectorChain calls the given list of CompositionSelectors in order.
+type CompositionSelectorChain struct {
+	list []CompositionSelector
+}
+
+// SelectComposition calls all SelectComposition functions of CompositionSelectors
+// in the list.
+func (r *CompositionSelectorChain) SelectComposition(ctx context.Context, cp resource.Composite) error {
+	for _, cs := range r.list {
+		if err := cs.SelectComposition(ctx, cp); err != nil {
```

Comment:
Wondering if we could return a reference here and then we would avoid calling all of the composition selectors passed in. We could then pass in the `APISelectorResolver` and if a selector is specified then we will never call the `DefaultCompositionSelector`. We also would not have to check for composition references or selectors in the defaulter because they would have already been used and the chain would have exited. What do you think?

Reply:
That would work for selector resolver and default selector but not much for enforced selector, where it overrides whatever is set as reference on the composite resource no matter what.

So, we're not only delegating the choice of reference but also the responsibility of setting the reference to the `CompositionSelector` so that individual implementations of `CompositionSelector`s could incorporate their choice about override and possibly other operations on the CR.

Reply:
Does this enforced selector exist already or are you just saying if it did then this model would not work? 

In my proposed model:
1. If you pass the `EnforcedSelector` first and a selector exists then it will always return a reference , and no other chained `CompositionSelector`s will be called.
2. If you pass the `EnforcedSelector` first and a selector does not exist then it will not return a reference, and other chained `CompositionSelector`s will be called.

In the current model:
1. If you pass the `EnforcedSelector` first and a selector exists then it will set the reference, and all other chained `CompositionSelector`s will be called.
2. If you pass the `EnforcedSelector` first and a selector does not exist then it will not modify the reference, and all other chained `CompositionSelector`s will be called.

I get what you mean when the `EnforcedSelector` is not passed first because then it will override any of the other `CompositionSelector`s, but that does not feel like a chain to me, it sounds more life modifiers. Generally with a chain I think of an ordered set of operations that are applied sequentially until one is satisfied. For instance, in some sort of authentication chain it will usually not check for a config file, find a valid one, then check for an environment variable after that and decide to toss out the config file if the environment variable exists (for example). In practice, the present `SelectorResolver`s will not run into this issue because of the way they check for a reference and selectors, but future override selectors that a user might define (such as something like `EnforceDefault`, which would override any references and selectors with the default) could cause surprising behavior when used with existing `CompositionSelector`s.

All of that being said, this is a UX thing that is not changing the actual functionality at this time, so if you feel strongly I would be willing to defer. It seems like the current model could have more complex behavior, but also be more powerful, while my proposal would be simpler to reason about but would limit the ability to use `CompositionSelector`s together in the same way.

Reply:
@muvaf note that I wrote this comment above before the update you just pushed, which I see includes `EnforcedCompositionSelector` :) I think my points still stand though if you want to weigh in further :+1:

Reply:
I think the difference between two models is about keeping some common logic in the `CompositionSelectorChain` object or not. I think we shouldn't return a reference in `Reconcile` object and set it, so, in order to accommodate that suggestion we'd need to end up with 2 interfaces; one for `CompositionSelectorChain` to be called by `Reconcile` and one for each of the selectors that are passed in.

As you said it's a UX thing and I think one interface and having the chain caller dumb seems simpler to me. Also, I pushed the enforced selector, too. You can see each selector has all of its logic in the function in an idempotent way so that the order of the selectors doesn't actually matter.

Reply:
> You can see each selector has all of its logic in the function in an idempotent way so that the order of the selectors doesn't actually matter.

Yep, I definitely appreciate that about these defined `CompositionSelector`s, my worry is the cognitive burden for someone adding a new `CompositionSelector` that may not be idempotent. However, maybe it is an okay expectation for us to have of users that they would know to write them in an idempotent way if they don't want consumers to worry about order. As I mentioned, I don't think there is a right way or a wrong way necessarily, just my personal preference. You have done more work on these controllers, so I will defer to your judgement :)